src/components/tl/ucp/reduce/reduce_dbt.c

```diff
@@ -0,0 +1,389 @@
+        return;
+    }
+
+    if (is_root && rank == t1.root) {
+        UCPCHECK_GOTO(ucc_mc_memcpy(args->dst.info.buffer, t1_rbuf,
+                                    data_size_t1, mtype, mtype), task, out);
+    }
+
+    if (is_root && rank == t2.root) {
+        UCPCHECK_GOTO(ucc_mc_memcpy(PTR_OFFSET(args->dst.info.buffer,
+                                               data_size_t1), t2_rbuf,
+                                               data_size_t2, mtype, mtype),
+                      task, out);
+    }
+
+    task->super.status = UCC_OK;
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_reduce_dbt_done", 0);
+out:
+    return;
+}
+
+ucc_status_t ucc_tl_ucp_reduce_dbt_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task       = ucc_derived_of(coll_task,
+                                                   ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team       = TASK_TEAM(task);
+    ucc_coll_args_t   *args       = &TASK_ARGS(task);
+    ucc_rank_t         rank       = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         team_size  = UCC_TL_TEAM_SIZE(team);
+    int                avg_pre_op =
+        UCC_TL_UCP_TEAM_LIB(TASK_TEAM(task))->cfg.reduce_avg_pre_op;
+    ucc_datatype_t     dt;
+    size_t             count, data_size;
+    ucc_status_t       status;
+
+    task->reduce_dbt.t1.recv           = 0;
+    task->reduce_dbt.t2.recv           = 0;
+    task->reduce_dbt.t1_reduction_comp = 0;
+    task->reduce_dbt.t2_reduction_comp = 0;
+    task->reduce_dbt.t1_send_comp      = 0;
+    task->reduce_dbt.t2_send_comp      = 0;
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    if (args->root == rank) {
+        count = args->dst.info.count;
+        dt    = args->dst.info.datatype;
+    } else {
+        count = args->src.info.count;
+        dt    = args->src.info.datatype;
+    }
+    data_size = count * ucc_dt_size(dt);
+
+    status = ucc_coll_task_get_executor(&task->super,
+                                        &task->reduce_dbt.executor);
+    if (ucc_unlikely(status != UCC_OK)) {
+        return status;
+    }
+
+    if (UCC_IS_INPLACE(*args) && (rank == args->root)) {
+        args->src.info.buffer = args->dst.info.buffer;
+    }
+
+    if (avg_pre_op && args->op == UCC_OP_AVG) {
+        /* In case of avg_pre_op, each process must divide itself by team_size */
+        status =
+            ucc_dt_reduce(args->src.info.buffer, args->src.info.buffer,
+                          PTR_OFFSET(task->reduce_dbt.scratch, data_size * 2),
+                          count, dt, args, UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA,
+                          1.0 / (double)(team_size * 2),
+                          task->reduce_dbt.executor, &task->reduce_dbt.etask);
+        if (ucc_unlikely(UCC_OK != status)) {
+            tl_error(UCC_TASK_LIB(task),
+                     "failed to perform dt reduction");
+            return status;
+        }
+        EXEC_TASK_WAIT(task->reduce_dbt.etask, status);
```

Comment:
It's ok, reduction here is local operation so it cannot cause deadlock and doesn't break ucc spec. However for GPU it might be not optimal since we limit concurrency. I would rather keep it as is for now, we can optimize it later.